packages/casting/src/follower-cosmjs.js

```diff
@@ -112,290 +145,442 @@ export const makeCosmjsFollower = (
+    );
+    return result.slice(dataPrefixBytes.length);
+  };
+
+  /**
+   * @param {number} height
+   */
+  const getUnprovenDataAtHeight = async height => {
+    const {
+      storeName,
+      storeSubkey,
+      dataPrefixBytes = defaultDataPrefixBytes,
+    } = await castingSpecP;
+
+    const values = await E(leader).mapEndpoints(where, async endpoint => {
+      const client = await provideTendermintClient(endpoint);
+      const response = await client.abciQuery({
+        path: `store/${storeName}/key`,
+        data: storeSubkey,
+        height,
+        prove: false,
       });
-      const { result, error } = collectSingle(values);
-      if (error !== null) {
-        throw error;
+      if (response.code !== 0) {
+        throw new Error(`Tendermint ABCI query failed: ${response.log}`);
       }
-      assert(result);
+      const { value } = response;
+      return value;
+    });
+    const result = collectSingle(values);
+    if (result.length === 0) {
+      // No data.
+      return result;
+    }
+
+    // Handle the data prefix if any.
+    assert(
+      result.length >= dataPrefixBytes.length,
+      X`result too short for data prefix ${dataPrefixBytes}`,
+    );
+    assert(
+      arrayEqual(dataPrefixBytes, result.subarray(0, dataPrefixBytes.length)),
+      X`${result} doesn't start with data prefix ${dataPrefixBytes}`,
+    );
+    return result.slice(dataPrefixBytes.length);
+  };
 
-      if (result.length === 0) {
-        // No data.
-        return result;
+  /**
+   * @param {number} blockHeight
+   */
+  const tryGetDataAtHeight = blockHeight => {
+    const getProvenData = () => getProvenDataAtHeight(blockHeight);
+    const getUnprovenData = () => getUnprovenDataAtHeight(blockHeight);
+    return queryVerifier(getProvenData, crash, getUnprovenData);
+  };
+
+  /**
+   * @param {number} blockHeight
+   */
+  const getDataAtHeight = async blockHeight => {
+    for (let attempt = 0; ; attempt += 1) {
+      try {
+        // AWAIT
+        return await tryGetDataAtHeight(blockHeight);
+      } catch (error) {
+        // We expect occasionally to see an error here if the chain has not
+        // reached the requested blockHeight.
+        await E(leader).retry(where, error, attempt);
+        continue;
       }
+    }
+  };
 
-      // Handle the data prefix if any.
-      assert(
-        result.length >= dataPrefixBytes.length,
-        X`result too short for data prefix ${dataPrefixBytes}`,
-      );
-      assert(
-        dataPrefixBytes.every((v, i) => v === result[i]),
-        X`${result} doesn't start with data prefix ${dataPrefixBytes}`,
-      );
-      return result.slice(dataPrefixBytes.length);
+  /**
+   * @param {number} blockHeight
+   * @param {Uint8Array} data
+   */
+  const streamCellForData = (blockHeight, data) => {
+    const cell = decode(data);
+
+    if (
+      cell &&
+      Array.isArray(cell.values) &&
+      typeof cell.blockHeight === 'string'
```

Comment:
```suggestion
      Array.isArray(cell.values) &&
      typeof cell.blockHeight === 'string' &&
      /^0$|^[1-9][0-9]*$/.test(cell.blockHeight)
```
Also, let's make an `isStreamCell` helper to avoid duplication and divergence.